glotaran/optimization/matrix_provider.py

```diff
@@ -0,0 +1,707 @@
+"""Module containing the matrix provider classes."""
+from __future__ import annotations
+
+import warnings
+from dataclasses import dataclass
+from dataclasses import replace
+
+import numba as nb
+import numpy as np
+import xarray as xr
+
+from glotaran.model import DatasetGroup
+from glotaran.model import DatasetModel
+from glotaran.model.interval_property import IntervalProperty
+from glotaran.optimization.data_provider import DataProvider
+from glotaran.optimization.data_provider import DataProviderLinked
+
+
+@dataclass
+class MatrixContainer:
+    """A container of matrix and the corresponding clp labels."""
+
+    clp_labels: list[str]
+    """The clp labels."""
+    matrix: np.ndarray
+    """The matrix."""
+
+    @staticmethod
+    @nb.jit(nopython=True, parallel=True)
+    def apply_weight(
+        matrix: np.typing.ArrayLike, weight: np.typing.ArrayLike
+    ) -> np.typing.ArrayLike:
+        """Apply weight on a matrix.
+
+        Parameters
+        ----------
+        matrix : np.typing.ArrayLike
+            The matrix.
+        weight : np.typing.ArrayLike
+            The weight.
+
+        Returns
+        -------
+        np.typing.ArrayLike
+            The weighted matrix.
+        """
+        matrix = matrix.copy()
+        for i in nb.prange(matrix.shape[1]):
+            matrix[:, i] *= weight
+        return matrix
```

Comment:
This is simple enough that we can use a pure numpy implementation which is actually faster

```python
In [1]: import numba as nb

In [2]: import numpy as np

In [3]: @nb.jit(nopython=True, parallel=True)
   ...: def apply_weight_nb(matrix, weight):
   ...:     matrix = matrix.copy()
   ...:     for i in nb.prange(matrix.shape[1]):
   ...:         matrix[:, i] *= weight
   ...:     return matrix
   ...:

In [4]: def apply_weight_np(matrix, weight):
   ...:     return (matrix.T * weight).T
   ...:

In [5]: matrix = np.ones((1000,200))

In [6]: weights = np.arange(1000)

In [7]: (apply_weight_nb(matrix,weights) == apply_weight_np(matrix,weights)).all()
Out[7]: True

In [8]: %timeit apply_weight_nb(matrix,weights)
749 µs ± 158 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [9]: %timeit apply_weight_np(matrix,weights)
134 µs ± 1.35 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [10]: %timeit apply_weight_nb(matrix,weights)
502 µs ± 20.5 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```